File: apps/android/domain/src/main/kotlin/com/sanvya/app/domain/csv/Csv.kt
package com.sanvya.app.domain.csv

/**
 * Minimal, dependency-free CSV (RFC-4180-ish): quotes, escaped quotes, CRLF.
 * Ported from apps/web/src/data/csv.ts, character for character.
 *
 * No CSV library on either platform, deliberately. Web hand-wrote this because
 * a dependency was not worth it; adopting a different parser on each phone
 * would give three implementations that disagree about the interesting cases --
 * an escaped quote inside a quoted cell, a bare CR, a trailing newline -- which
 * are exactly the cases a real bank export contains.
 *
 * `downloadText` is NOT ported: saving a file is a platform concern, and there
 * is nothing shared about a browser anchor click, an Android SAF intent and a
 * UIActivityViewController.
 *
 * Mirrors apps/ios/Domain/Sources/Domain/Csv.swift.
 */

/** Parse CSV text into rows of string cells. Auto-detects `,` vs `;`. */
fun parseCsv(text: String, delimiter: String? = null): List<List<String>> {
    val src = text.removePrefix("﻿") // strip BOM
    val delim = (delimiter ?: detectDelimiter(src)).first()
    val rows = mutableListOf<List<String>>()
    var row = mutableListOf<String>()
    val cell = StringBuilder()
    var inQuotes = false
    var i = 0
    while (i < src.length) {
        val ch = src[i]
        when {
            inQuotes -> when {
                ch == '"' && i + 1 < src.length && src[i + 1] == '"' -> {
                    cell.append('"'); i++
                }
                ch == '"' -> inQuotes = false
                else -> cell.append(ch)
            }
            ch == '"' -> inQuotes = true
            ch == delim -> { row.add(cell.toString()); cell.setLength(0) }
            ch == '\n' -> {
                row.add(cell.toString()); rows.add(row); row = mutableListOf(); cell.setLength(0)
            }
            // A bare CR is dropped; the LF branch above ends the row. Web's
            // comment says "handled by the \n branch", which is only true for
            // CRLF -- a lone CR line ending produces one very long row on all
            // three platforms. Preserved, because a file that old would import
            // identically wrong everywhere rather than differently.
            ch == '\r' -> Unit
            else -> cell.append(ch)
        }
        i++
    }
    if (cell.isNotEmpty() || row.isNotEmpty()) { row.add(cell.toString()); rows.add(row) }
    return rows.filter { r -> r.any { it.trim().isNotEmpty() } }
}

private fun detectDelimiter(text: String): String {
    val newline = text.indexOf('\n')
    val firstLine = if (newline >= 0) text.substring(0, newline) else text
    val commas = firstLine.count { it == ',' }
    val semis = firstLine.count { it == ';' }
    return if (semis > commas) ";" else ","
}

/** Parse into header-keyed records (headers lowercased + trimmed). */
fun parseRecords(text: String, delimiter: String? = null): List<Map<String, String>> {
    val rows = parseCsv(text, delimiter)
    if (rows.isEmpty()) return emptyList()
    val headers = rows[0].map { it.trim().lowercase() }
    return rows.drop(1).map { r ->
        // A LinkedHashMap so the record's iteration order is the file's column
        // order -- nothing depends on it today, but a vector compares JSON.
        val rec = LinkedHashMap<String, String>()
        headers.forEachIndexed { index, h -> rec[h] = (r.getOrNull(index) ?: "").trim() }
        rec
    }
}

/** Serialize rows (first row = header) to CSV text. */
fun toCsv(rows: List<List<String?>>): String =
    rows.joinToString("\r\n") { r -> r.joinToString(",") { escapeCell(it) } }

private fun escapeCell(v: String?): String {
    val s = v ?: ""
    return if (s.any { it == '"' || it == ',' || it == '\r' || it == '\n' }) {
        "\"" + s.replace("\"", "\"\"") + "\""
    } else {
        s
    }
}

/**
 * A minor-unit amount as the plain major-unit number a CSV cell holds:
 * `49900` INR -> `"499.00"`, `500` JPY -> `"500"`.
 *
 * Web writes `(amount / 100).toFixed(2)` and its importer reads that straight
 * back, so a JPY 500 charge exports as "5.00" and re-imports as JPY 5 -- a
 * round trip that loses money for every currency that is not two-decimal. Using
 * the currency's own minor-unit count in both directions makes the round trip
 * lossless, and for INR, USD and EUR produces byte-identical output to web's.
 *
 * Deliberately NOT `formatMoney`: this is a machine-readable cell. A grouped,
 * symbol-prefixed, locale-formatted number is what the importer's amount parser
 * has to fight its way back out of.
 */
fun majorText(minor: Long, currency: String): String {
    val decimals = com.sanvya.app.domain.money.minorUnits(currency)
    val major = com.sanvya.app.domain.money.toMajor(
        com.sanvya.app.domain.money.money(minor, currency)
    )
    return String.format(java.util.Locale.ROOT, "%.${decimals}f", major)
}

File: apps/ios/Domain/Sources/Domain/Csv.swift
import Foundation

/// Minimal, dependency-free CSV (RFC-4180-ish): quotes, escaped quotes, CRLF.
/// Ported from apps/web/src/data/csv.ts, character for character.
///
/// No CSV library on either platform, deliberately. Web hand-wrote this because
/// a dependency was not worth it; adopting a different parser on each phone
/// would give three implementations that disagree about the interesting cases —
/// an escaped quote inside a quoted cell, a bare CR, a trailing newline — which
/// are exactly the cases a real bank export contains.
///
/// `downloadText` is NOT ported: saving a file is a platform concern, and there
/// is nothing shared about a browser anchor click, an Android SAF intent and a
/// `UIActivityViewController`.
///
/// Mirrors apps/android/domain/.../csv/Csv.kt.

/// Parse CSV text into rows of string cells. Auto-detects `,` vs `;`.
public func parseCsv(_ text: String, delimiter: String? = nil) -> [[String]] {
    let src = text.hasPrefix("\u{FEFF}") ? String(text.dropFirst()) : text // strip BOM
    let delim = Character(delimiter ?? detectDelimiter(src))
    var rows: [[String]] = []
    var row: [String] = []
    var cell = ""
    var inQuotes = false

    let chars = Array(src)
    var i = 0
    while i < chars.count {
        let ch = chars[i]
        if inQuotes {
            if ch == "\"" {
                if i + 1 < chars.count && chars[i + 1] == "\"" {
                    cell.append("\"")
                    i += 1
                } else {
                    inQuotes = false
                }
            } else {
                cell.append(ch)
            }
        } else if ch == "\"" {
            inQuotes = true
        } else if ch == delim {
            row.append(cell)
            cell = ""
        } else if ch == "\n" {
            row.append(cell)
            rows.append(row)
            row = []
            cell = ""
        } else if ch == "\r" {
            // A bare CR is dropped; the LF branch above ends the row. Web's
            // comment says "handled by the \n branch", which is only true for
            // CRLF — a lone CR line ending produces one very long row on all
            // three platforms. Preserved, because a file that old would import
            // identically wrong everywhere rather than differently.
        } else {
            cell.append(ch)
        }
        i += 1
    }
    if !cell.isEmpty || !row.isEmpty {
        row.append(cell)
        rows.append(row)
    }
    return rows.filter { r in r.contains { !$0.trimmingCharacters(in: .whitespaces).isEmpty } }
}

private func detectDelimiter(_ text: String) -> String {
    let firstLine = text.prefix { $0 != "\n" }
    let commas = firstLine.filter { $0 == "," }.count
    let semis = firstLine.filter { $0 == ";" }.count
    return semis > commas ? ";" : ","
}

/// Parse into header-keyed records (headers lowercased + trimmed).
///
/// The key ORDER is the file's column order — a plain dictionary would lose it,
/// and a vector compares serialised JSON.
public struct CsvRecord: Sendable {
    public let keys: [String]
    private let values: [String: String]

    public init(keys: [String], values: [String: String]) {
        self.keys = keys
        self.values = values
    }

    public subscript(key: String) -> String? { values[key] }
}

public func parseRecords(_ text: String, delimiter: String? = nil) -> [CsvRecord] {
    let rows = parseCsv(text, delimiter: delimiter)
    guard let headerRow = rows.first else { return [] }
    let headers = headerRow.map { $0.trimmingCharacters(in: .whitespaces).lowercased() }
    return rows.dropFirst().map { r in
        var values: [String: String] = [:]
        for (index, h) in headers.enumerated() {
            values[h] = (index < r.count ? r[index] : "").trimmingCharacters(in: .whitespaces)
        }
        return CsvRecord(keys: headers, values: values)
    }
}

/// Serialize rows (first row = header) to CSV text.
public func toCsv(_ rows: [[String?]]) -> String {
    rows.map { r in r.map(escapeCell).joined(separator: ",") }.joined(separator: "\r\n")
}

private func escapeCell(_ v: String?) -> String {
    let s = v ?? ""
    let needsQuotes = s.contains { $0 == "\"" || $0 == "," || $0 == "\r" || $0 == "\n" }
    return needsQuotes ? "\"" + s.replacingOccurrences(of: "\"", with: "\"\"") + "\"" : s
}

/// JavaScript's `Number.parseFloat`, which reads a LEADING numeric prefix and
/// ignores the rest. Swift's `Double(_:)` requires the whole string, so
/// `"1.2.3"` is 1.2 in the browser and nil here — and a real bank export does
/// contain cells like that.
func jsParseFloat(_ s: String) -> Double? {
    let pattern = "^[+-]?(?:\\d+(?:\\.\\d*)?|\\.\\d+)(?:[eE][+-]?\\d+)?"
    guard let regex = try? NSRegularExpression(pattern: pattern),
          let match = regex.firstMatch(in: s, range: NSRange(s.startIndex..., in: s)),
          let range = Range(match.range, in: s) else { return nil }
    return Double(s[range])
}
